### ISSUE TITLE
Generalize the "notes" part of `citar-open`

### ISSUE THREAD

**Issue by bdarcus** [emacs-citar/citar#612]
We need to adjust `citar-open` to #603 (or drop the command altogether?).

The problem currently is the candidate list here is file paths, and typed as such for Embark.

And `citar--select-resource` takes a list of files as input.

But for org-roam v2, for example, with its more general concept of note, the candidates would be "org-roam-ref" "nodes".

I need to look, but I suspect we need to add another defcustom for the candidates.

-------------

## Notes

- `org-roam-ref-read` is the completion function
- it has a customizable annotation function, to display the metadata
- it uses `org-roam-ref-read--completions`; its candidates are a key and a list of org-roam nodes, which look like this:

```elisp
(#("agnew2003" 0 9
   (node #s(org-roam-node :file "/home/bruce/org/roam/biblio/agnew2003.org" :file-title nil :file-hash nil :file-atime nil :file-mtime nil :id "fd1e8dfa-8bfd-43fb-884a-b350a7a0850a" :level nil :point 1 :todo nil :priority nil :scheduled nil :deadline nil :title "Notes on Agnew, From Megalopolis to Global City-Region?" :properties nil :olp nil :tags nil :aliases nil :refs nil)
          type "cite"))
  . #s(org-roam-node :file "/home/bruce/org/roam/biblio/agnew2003.org" :file-title nil :file-hash nil :file-atime nil :file-mtime nil :id "fd1e8dfa-8bfd-43fb-884a-b350a7a0850a" :level nil :point 1 :todo nil :priority nil :scheduled nil :deadline nil :title "Notes on Agnew, From Megalopolis to Global City-Region?" :properties nil :olp nil :tags nil :aliases nil :refs nil))
```

**Comment by aikrahguzar** [comment]
> @aikrahguzar no rush, but since you've been working on and thinking about `completing-read` and embark, any thoughts on how to handle this elegantly?

I am not much of a note taker :)

From what I can tell you want to be able to address both files or location within a file and what springs to mind is bookmarks. I haven't used them myself but from what I have read about them they should work for both cases i.e. convert both filenames and org-roam nodes to a bookmark and use bookmark as an embark category. 

An other approach would be provide something basic here and have a separate package which depends on `consult` to provide a richer interface. `consult--multi` can accept multiple sources  with different categories and `embark-act` will work with the correct category of the candidate. Try `M-x consult-buffer` and it has both buffers and files in there and `embark-act` picks the right category. So you might want to look at how it achieves this if you prefer files and org-roam nodes to have different categories. 

But I would say if you are really pushing the boundaries of `completing-read` it is better to just depend on consult since it has mastered that art. The only downside of that approach is that one has to depend on private functions in consult.

**Comment by bdarcus** [comment]
> From what I can tell you want to be able to address both files or location within a file and what springs to mind is bookmarks. I haven't used them myself but from what I have read about them they should work for both cases i.e. convert both filenames and org-roam nodes to a bookmark and use bookmark as an embark category.

Ah, I hadn't thought about that.
 
> An other approach would be provide something basic here and have a separate package which depends on `consult` to provide a richer interface. `consult--multi` can accept multiple sources with different categories and `embark-act` will work with the correct category of the candidate.

I had earlier thought of this, if not specifically for this issue. But that is certainly a possibility, in which case I might just drop `citar-open`.

I'll need to experiment with `consult--multi` again.

@mclearc is working on this `consult-notes` package, which makes use of `consult--multi`.

https://github.com/mclear-tools/consult-notes

It's purely focused on notes, however, whereas `citar-open` is focused on resources related to citation references, including notes.

**Comment by bdarcus** [comment]
@mclearc

> It's purely focused on notes, however, whereas `citar-open` is focused on resources related to citation references, including notes.

Maybe a wild idea:

Do you think it would be feasible to replace `citar-open` with something based on `consult--multi`, where one could plug in different associated notes functions, including one from `consult-notes`??

I played again with `consult--multi`, but I couldn't figure out the `org-roam` part; e.g this is broken, but illustrates the idea:

```elisp
(defun concite-notes-org-roam ()
  "Notes."
  (org-roam-ref-read--completions))

(defvar concite-source-notes
  `(:name "Notes"
    :narrow ?n
    :category org-roam-node
    :items ,#'concite-notes-org-roam))
```

**Comment by mclearc** [comment]
@bdarcus -- yeah interesting idea. I'll see if i can get something working.

**Comment by mclearc** [comment]
there is [this function](https://github.com/jgru/consult-org-roam/blob/05cec288f931a1f3cd5984b88a79f1339110e4b9/consult-org-roam.el#L175) from the [consult-org-roam](https://github.com/jgru/consult-org-roam) package that might be useful. I'm not sure I have time today to get into this all that much but we'll see...

**Comment by bdarcus** [comment]
No rush at all; it (whether to remove citar-open from here) is just
the kind of question I want to settle before I tag 1.0, which is
likely sometime this month.

**Comment by bdarcus** [comment]
I figured it out:

I think the constraint that `consult--multi` imposes is that candidates must be a list of strings. More complex candidates as in org-roam or citar are not allowed.

But one can probably propertize those strings to embed additional metadata in them.

And the other thing I don't see ATM: affixation functions.

**Comment by mclearc** [comment]
Cool -- i've generalized the `consult-notes` function so that it accepts arbitrary sources as normal for `consult--multi` so this works. I'm pretty sure I can figure out how to add annotation data of some sort. 

![Screen Shot 2022-06-03 at 17 43 51](https://user-images.githubusercontent.com/1502422/171957856-fd4b7e63-7e3e-4991-87b2-da0b73cdd164.png)

**Comment by mclearc** [comment]
> I also am not a fan of those completion candidates; completing on the keys.

This seems better:
<img width="855" alt="Screen Shot 2022-06-04 at 14 48 43" src="https://user-images.githubusercontent.com/1502422/172021514-965a74d6-9428-491b-b3ee-a21d62433a50.png">

the relevant code being:
```emacs-lisp
    :items ,(lambda () (let* ((node (mapcar #'cdr (org-roam-ref-read--completions)))
                                   (title (mapcar #'org-roam-node-title node)))
                    (progn title)))
```
I'm having real trouble (perhaps predictably) getting any good annotations though. We'll see.

**Comment by mclearc** [comment]
Actually -- I think this works:

<img width="1046" alt="Screen Shot 2022-06-04 at 15 15 39" src="https://user-images.githubusercontent.com/1502422/172022423-a7d7a360-6942-4327-80d9-f575fc30a252.png">

```emacs-lisp
    :annotate ,(lambda (cand)
                 (let* ((file (org-roam-node-file (org-roam-node-from-title-or-alias cand)))
                        (attrs (file-attributes file))
                        (dir (file-name-nondirectory (directory-file-name (file-name-directory file))))
                        (size
                         (file-size-human-readable (file-attribute-size (file-attributes file))))
                        (time (consult-notes--time (file-attribute-modification-time attrs)))
                        )
                   (concat (format "%7s %7s" dir size) "  " (format "%10s" time))
                   ))
```

**Comment by bdarcus** [comment]
Closed via #634
